src/index.spec.ts

```diff
@@ -1,12 +1,182 @@
+      {
+        name: 'true evaluating a negative number',
+        actual: -123,
+        expected: true,
+      },
+      { name: 'false evaluating a string', actual: '1', expected: false },
+      { name: 'false evaluating a float', actual: 1.25, expected: false },
+    ],
+  );
+});
+
+describe('toLowerCase', () => {
+  cases(
+    `should return`,
+    (opts) => {
+      expect(toLowerCase(opts.actual)).toEqual(opts.expected);
+    },
+    [
+      {
+        name: 'prueba when PrueBa is given',
+        actual: 'PrueBa',
+        expected: 'prueba',
+      },
+      {
+        name: 'the same string when a lowercase string is given',
+        actual: 'lowercase',
+        expected: 'lowercase',
+      },
+      {
+        name: 'return error when passing null or an empty string',
+        actual: null,
+        expected: 'Please provide a string',
+      },
+    ],
+  );
+});
+
+describe('removeDuplicatesFromArray', () => {
+  cases(
+    `should return`,
+    (opts) => {
+      expect(removeDuplicatesFromArray(opts.actual)).toStrictEqual(
+        opts.expected,
+      );
+    },
+    [
+      {
+        name: '[1,2,3] when [1,2,2,3,1] is evaluated',
+        actual: [1, 2, 2, 3, 1],
+        expected: [1, 2, 3],
+      },
+      {
+        name: `['1','2','3'] when ['1','2','2','3','1'] is evaluated`,
+        actual: ['1', '2', '2', '3', '1'],
+        expected: ['1', '2', '3'],
+      },
+      {
+        name: 'the same array when the array only has one element',
+        actual: ['one'],
+        expected: ['one'],
+      },
+    ],
+  );
+
+  cases(
+    'should throw an error',
+    (opts) => {
+      expect(() => removeDuplicatesFromArray(opts.actual as any)).toThrowError(
+        opts.expected,
+      );
+    },
+    [
+      {
+        name: 'when there is no an array given',
+        actual: {},
+        expected: 'please provide an array of numbers or strings',
+      },
+    ],
+  );
+});
+
+describe('createRandomProduct', () => {
+  it('should return a random product when giving a creator role email', () => {
+    const creatorEmail = 'clark@kent.com';
+    const actual = createRandomProduct(creatorEmail);
+    expect(actual).not.toBeNull();
+  });
+
+  it('should throw an error when user does not have creator role', () => {
+    const noCreatorUser = 'diana@themyscira.com';
+    expect(() => createRandomProduct(noCreatorUser)).toThrowError(
+      'You are not allowed to create products',
+    );
+  });
+
+  it('should throw an error when passing an invalid email', () => {
+    const invalidEmail = 'invalidemail@fake.com';
+    expect(() =>
+      createRandomProduct(invalidEmail),
+    ).toThrowErrorMatchingInlineSnapshot(
+      `"Cannot read properties of undefined (reading 'role')"`,
+    );
+  });
+});
+
+describe('createProduct', () => {
+  const testProduct = {
+    name: 'motorcycle',
+    description: 'arch motorcycle',
+    price: 1000,
+    tags: ['at least one'],
+  };
```

Comment:
In this case, if you are defining a variable to use in this code block is better if you define it into `beforeAll(() => {...})` and if you need to reset the values in every test you can use beforeEach(() => {...}) and you need to declare the variable outside of the `beforeAll` or `beforeEach` with his type

```
describe('createProduct', () => {
    let testProduct : {
        name: string,
        description: string,
        price: number,
        tags: string[]
    }

    beforeAll(()=>{
       testProduct = {
             name: 'motorcycle',
            description: 'arch motorcycle',
            price: 1000,
            tags: ['at least one'],
       };       
   })
   ... 
})